### ISSUE TITLE
Add support for GPU_Query_Support

### ISSUE THREAD

**Issue by PhilipVinc** [mpi4py/mpi4py#469]
Hello, I'm the developer of [mpi4jax](https://github.com/mpi4jax/mpi4jax). 
I'd like to thank you for all your effort in wrapping MPI within Python that made our life already much easier when trying to wrap MPI in Jax.

In order to support GPU-direct communications in a very simple, automatic way, we would love to be able to query whether it is supported natively by the MPI library.

I recently found out that at least MPICH supports a (maybe non standard?) way to query this for arbitrary backends
https://www.mpich.org/static/docs/latest/www3/MPIX_GPU_query_support.html
and that OpenMPI supports a version for CUDA
https://www.open-mpi.org/doc/v4.1/man3/MPIX_Query_cuda_support.3.php

Would it be possible to expose this from mpi4py?

**Comment by leofang** [comment]
We won't support a non standard query because none of the MPI implementations was willing to work with us (we asked this feature years ago), and it is too much burden on our side. However, with MPI 4.1 there will be an alternative: https://github.com/mpi4py/mpi4py/issues/162#issuecomment-1826133459. (btw this issue is a duplicate of that.)

**Comment by PhilipVinc** [comment]
So if I understand correctly with MPI 4.1 there will be `mpi_memory_alloc_kinds` which may be used by implementations to advertise this kind of information.

Will mpi4py support `mpi_memory_alloc_kinds` in a future release?

**Comment by leofang** [comment]
IIUC support for MPI 4.1 just got released with MPICH 4.2.0; I lost track of progress on the Open MPI side, maybe @dalcinl has better idea.

**Comment by dalcinl** [comment]
> Will mpi4py support `mpi_memory_alloc_kinds` in a future release? I can't find any reference within mpi4py

No need to wait for a future release... The current release supports all the `Info` APIs, therefore you can use `mpi_memory_alloc_kinds` right now provided the backend MPI implementation supports it.

> I lost track of progress on the Open MPI side

Open MPI has partial support for MPI 4.0, it is currently missing all of the large-count variants. AFAIK, they have not added any of the new MPI 4.1 routines, but I may be wrong. However, `mpi_memory_alloc_kinds` would be something Open MPI could support easily, as it does not imply any API/ABI changes, not even new MPI routines.

**Comment by PhilipVinc** [comment]
I just tried reinstalling `mpich==4.2` and mpi4py but I don't understand how to use the `MPI.Info` objects to query this information?

**Comment by dalcinl** [comment]
`Info` objects are akin to Python `dict` objects. Moreover, in Python, then can be used with a dict-like interface. You will have to search the MPI standard document to learn about the `mpi_memory_alloc_kinds` key. You should start by looking for the key and its value in the `INFO_ENV` object. In case of trouble or doubt, the answers are in the MPICH side.

**Comment by leofang** [comment]
As a digression, for my own curiosity @PhilipVinc may I ask why mpi4jax is needed? Jax now supports DLPack and CUDA Array Interface, and mpi4py is supposed to work with either protocols out of box. Are we missing something?

**Comment by PhilipVinc** [comment]
Those are just 'ease of life things' but mpi4py operations are foreign to jax's transformations.

When jax compiles some function, he does so by tracing your function, and you can only call primitives that jax knows how to handle and tell jax how those primitives operate on the data and transform when using batching transforms`jax.vmap`, or autodiff transforms`jax.vjp` and over functions.

Jax authors have written a small set of primitives, such as tensor multiplication, reduce, sum... etc... and they have reimplemented numpy on top of those, we had to do the same for MPI.

Moreover, Jax's compiler, XLA, requires that the primitives be C functions with a very specific signature.

So, in order to really write 'differentiable programs' where you can compute gradients of code that performs MPI operations, you need to be able to expose MPI operations as primitives, and that's what we do in mpi4jax.
For example, in the case of Allreduce, we have to [declare the 'abstract rule'](https://github.com/mpi4jax/mpi4jax/blob/b7d19a041759ad85a2b04e010867920ed915032f/mpi4jax/_src/collective_ops/allreduce.py#L175) that tells jax how the output shape is related to the input shape, the [batching rule](https://github.com/mpi4jax/mpi4jax/blob/b7d19a041759ad85a2b04e010867920ed915032f/mpi4jax/_src/collective_ops/allreduce.py#L182), telling jax how to transform with `jax.vmap` a call to all reduce (well, this is trivial, but he needs to be told), how to compute [the jacobian vector product](https://github.com/mpi4jax/mpi4jax/blob/b7d19a041759ad85a2b04e010867920ed915032f/mpi4jax/_src/collective_ops/allreduce.py#L188) for computing the forward mode gradient, and finally how to compile the [cpu](https://github.com/mpi4jax/mpi4jax/blob/b7d19a041759ad85a2b04e010867920ed915032f/mpi4jax/_src/collective_ops/allreduce.py#L83) and gpu versions of the operation.

It's a lot of boilerplate, while the code is relatively straightforward, but it allows for using MPI at zero cost.

Funnily enough, those days jax people have built into jax some similar stuff (actually, much more elegant) using gloo, but MPI is still at least one order of magnitude faster...
